logic/source_map.go

```diff
@@ -0,0 +1,120 @@
+package logic
+
+import "strings"
+
+// SourceMap provides a mapping of the source to assembled program
+type SourceMap struct {
+	Version    int      `json:"version"`
+	File       string   `json:"file,omitempty"`
+	SourceRoot string   `json:"sourceRoot,omitempty"`
+	Sources    []string `json:"sources"`
+	Names      []string `json:"names"`
+	Mapping    string   `json:"mapping"`
+	// Decoded mapping results
+	LineToPc map[int][]int
+	PcToLine map[int]int
+}
+
+func NewSourceMap(ism map[string]interface{}) SourceMap {
+	sm := SourceMap{}
+
+	if v, ok := ism["version"]; ok {
+		sm.Version = int(v.(float64))
+	}
+
+	if f, ok := ism["file"]; ok {
+		sm.File = f.(string)
+	}
+
+	if sr, ok := ism["sourceRoot"]; ok {
+		sm.SourceRoot = sr.(string)
+	}
+
+	if srcs, ok := ism["sources"]; ok {
+		srcSlice := srcs.([]interface{})
+		sm.Sources = make([]string, len(srcSlice))
+		for idx, s := range srcSlice {
+			sm.Sources[idx] = s.(string)
+		}
+	}
+
+	if names, ok := ism["names"]; ok {
+		nameSlice := names.([]interface{})
+		sm.Names = make([]string, len(nameSlice))
+		for idx, n := range nameSlice {
+			sm.Names[idx] = n.(string)
+		}
+	}
+
+	if m, ok := ism["mapping"]; ok {
+		sm.Mapping = m.(string)
+	}
+
+	sm.PcToLine = map[int]int{0: 0}
+	sm.LineToPc = map[int][]int{0: {0}}
+
+	lastLine := 0
+	for idx, chunk := range strings.Split(sm.Mapping, ";") {
+		vals := DecodeSourceMapLine(chunk)
+		if len(vals) > 3 {
+			lineNum := vals[2]
+			if _, ok := sm.LineToPc[lineNum]; !ok {
+				sm.LineToPc[lineNum] = []int{}
+			}
+			sm.LineToPc[lineNum] = append(sm.LineToPc[lineNum], idx)
+
+			lastLine = lineNum
+		}
+		sm.PcToLine[idx] = lastLine
+	}
+
+	return sm
+}
+
+func (s *SourceMap) GetLineForPc(pc int) int {
+	return s.PcToLine[pc]
+}
+
+func (s *SourceMap) GetPCsForLine(line int) []int {
```

Comment:
@barnjamin Minor and optional:  Elsewhere, the PR does _not_ capitalize _PC_.  Propose lowercase for consistency.

```suggestion
func (s *SourceMap) GetPcsForLine(line int) []int {
```